smartsim/_core/launcher/dragon/pqueue.py

```diff
@@ -0,0 +1,464 @@
+
+        return self._nodes[host]
+
+    def decrement(
+        self, host: str, tracking_id: t.Optional[str] = None
+    ) -> NodeReferenceCount:
+        """Directly decrement the reference count of a given node and ensure the
+        ref counter is marked as dirty to trigger a reordering
+
+        :param host: a hostname that should have a reference counter decremented
+        :param tracking_id: unique task identifier to remove"""
+        with self._lock:
+            tracked_node = self._nodes[host]
+            tracked_node.remove(tracking_id)
+
+            return tracked_node
+
+    def _create_sub_heap(
+        self,
+        hosts: t.Optional[t.List[str]] = None,
+        filter_on: t.Optional[PrioritizerFilter] = None,
+    ) -> t.List[_TrackedNode]:
+        """Create a new heap from the primary heap with user-specified nodes
+
+        :param hosts: a list of hostnames used to filter the available nodes
+        :returns: a list of assigned reference counters
+        """
+        nodes_tracking_info: t.List[_TrackedNode] = []
+        heap = self._get_filtered_heap(filter_on)
+
+        # Collect all the tracking info for the requested nodes...
+        for node in heap:
+            if not hosts or node.hostname in hosts:
+                nodes_tracking_info.append(node)
+
+        # ... and use it to create a new heap from a specified subset of nodes
+        heapq.heapify(nodes_tracking_info)
+
+        return nodes_tracking_info
+
+    def unassigned(
+        self, heap: t.Optional[t.List[_TrackedNode]] = None
+    ) -> t.Sequence[Node]:
+        """Select nodes that are currently not assigned a task
+
+        :param heap: a subset of the node heap to consider
+        :returns: a list of reference counts for all unassigned nodes"""
+        if heap is None:
+            heap = list(self._nodes.values())
+
+        nodes: t.List[_TrackedNode] = []
+        for item in heap:
+            if item.num_refs == 0:
+                nodes.append(item)
+        return nodes
+
+    def assigned(
+        self, heap: t.Optional[t.List[_TrackedNode]] = None
+    ) -> t.Sequence[Node]:
+        """Helper method to identify the nodes that are currently assigned
+
+        :param heap: a subset of the node heap to consider
+        :returns: a list of reference counts for all assigned nodes"""
+        if heap is None:
+            heap = list(self._nodes.values())
+
+        nodes: t.List[_TrackedNode] = []
+        for item in heap:
+            if item.num_refs > 0:
+                nodes.append(item)
+        return nodes
+
+    def _check_satisfiable_n(
+        self, num_items: int, heap: t.Optional[t.List[_TrackedNode]] = None
+    ) -> bool:
+        """Validates that a request for some number of nodes `n` can be
+        satisfied by the prioritizer given the set of nodes available
+
+        :param num_items: the desired number of nodes to allocate
+        :param heap: a subset of the node heap to consider
```

Comment:
After seeing this line, it made me wonder if in this class "heap" is misused in a couple of places.  The type of this argument is optional ``List`` which isn't a heap... is there a better nomenclature for instances where we aren't dealing with a heap but just a set/list of nodes?